src/documents/tests/test_consumer.py

```diff
@@ -17,6 +30,33 @@ def test__get_parser_class_1_parser(self):
             self.DummyParser
         )
 
+    def test_open_errors(self):
+        """
+        Test that exceptions on open do not crash the consumer
+
+        It is not enough to test the consumption process with `open` throwing
+        an error everytime.  The first open call could succeed, but then the
+        file gets moved or permission changed by another process (e.g. user
+        manually intervening or a second paperless process).  Thus we need to
+        check what happens when only the n-th call fails.
+        """
+        
+        # open can throw PermissionError or FileNotFoundError
+        for error in [PermissionError(), FileNotFoundError()]:
+            # Check exception on n-th open call
+            for n in range(0, 4):
+                # Properly mock open
+                mockfile = mock.Mock()
+                mockfile_context_manager = mock.MagicMock()
+                # Return "I'm a pdf" on read()
+                mockfile_context_manager.read.return_value = StringIO("I'm a pdf")
+                mockfile.__enter__ = mockfile_context_manager
+                mockfile.__exit__ = mock.MagicMock()
+                with mock.patch("builtins.open", side_effect=error_on_call_n(error, n)) as mockopen:
```

Comment:
If I understand this correctly, you've counted the number of times we use `open()` in the consumer path and are looping over each making sure we handle `PermissionError` and `FileNotFoundError`.  Given that it's entirely likely that future iterations of the code may move these uses of `open()` around, and even add/remove some, I would think that this makes the test unnecessarily brittle.  Wouldn't it be better to write separate unit tests for each method with failing uses of `open()` expressly defined?  At least then if we change things, it's easy to see why a test failed and fix it.

Reply:
I think more of this testcase as an integration test, so that `consume()` is tested as close as possible to production as possible. Optimally it should never silently fail. When unittesting/testing subfunctions separately it is a) similar easy to forget to test a subfunction b) disregarding interactions between subfunctions. But I agree that this current design is brittle.

I see four ways forward:

1) Go full integration testing, run paperless in some kind of VM and let it actually consume stuff. Pro: Closest to production deployment Cons: Might be generally hard (syncronizing failing conditions, running in vm, proper sampling of input and conditions)
2) Keep the current approach but try to make it less brittle, increase the number of opens tested, or maybe automatically count how often open is called. Add similar functions that might fail. Pros: If done correctly should be faster and easier than 1, but have similar rates of catching bugs. Cons: Brittle, limited sampling of failing conditions, might produce weird failing messages.
3) Go more direction unittesting, test seperate functions. Pros: Should be less brittle (regarding refactoring) Cons: Does not capture interactions between units, is "far away from production conditions".
4) Cover consume() in a `try... catch` and display any such error messages prominently to the user. Make it easy for the user to a) see that one document failed consumption (no silent dropping of documents) b) post error messages to the issue tracker. Wasn't there a discussion on automatically sending logs somewhere? Pros: Captures all bugs in production. User gets feedback when something goes wrong, in the other cases not identified bugs could still lead to silently dropping documents. Cons: Captures bugs only in production, does not actually identify bugs when developing.

When I wrote the PR I was favoring 2. I've sitting a different version of the PR in a local repository, where I rewrote the tests to use pytests advanced features. But I actually couldn't reproduce some crashes which I thought should be there: `open` throwing errors did not crash consume(), but I couldn't identify why.

Currently I'd probably go for 4.